### ISSUE TITLE
busybox modprobe incompatibility

### ISSUE THREAD

**Issue by Geertsky** [dracutdevs/dracut#2581]
**Describe the bug**
Including busybox causes the emergency shell to get started due to incompatibility of` modprobe`.

**Distribution used**
Fedora38

**Dracut version**
059-5.fc38

**Init system**
systemd

**To Reproduce**
```
sudo dracut -NM -a "sshd busybox network lvm systemd-resolved" <path to initramfs> $(uname -r)
```

**Expected behavior**
Normal bootup

**Additional context**
It seems a problem with dropbox modprobe not supporting the `--all` flag. See `rdsosreport.txt` line `1099`

**Comment by Geertsky** [comment]
Tried to exclude the iscsi module but the problem persists...

```
sudo dracut -NM -m bash systemd systemd-initrd systemd-resolved systemd-sysusers modsign busybox nss-softokn dbus-broker dbus i18n network-manager network ifcfg sshd drm plymouth bcache btrfs crypt dm dmraid kernel-modules kernel-modules-extra kernel-network-modules lvm mdraid nvdimm qemu qemu-net cifs  lunmask nfs resume rootfs-block terminfo udev-rules virtiofs dracut-systemd usrmount base fs-lib memstrack shutdown /tmp/initramfs-test.img 6.6.3-100.fc38.x86_64 --force
```
[rdsosreport.txt](https://github.com/dracutdevs/dracut/files/13625426/rdsosreport.txt)

**Comment by Geertsky** [comment]
@LaszloGombos Seems to work. I'm still having problems with `busybox mount` not being able to mount `xfs` or `vfat` filesystems though...

**Comment by Geertsky** [comment]
@LaszloGombos Sorry for the late reply... But you're right! That did it! Wondering why dracut with busybox doesn't autoload these modules though...

**Comment by dbnicholson** [comment]
It would be nice if you could provide an allowlist or denylist for the busybox programs. We recently ran into failures because the busybox `blkid` implementation is a toy that's nothing like the real thing. Something like `busyboxapplets` and `omit_busyboxapplets`. Would something like that be acceptable?

**Comment by dbnicholson** [comment]
> @dbnicholson In most cases blkid can be replaced by udevadm and dracut can be more portable with changes like this - #2033 .

In some cases, sure. In our case we have a dracut module that expands the root partition. In order to know what type of partitioning is needed, we use `blkid -o value -s PTTYPE`. We could switch to `lsblk` or similar, but that depends on udev setting the properties. That introduces a potential failure if that hasn't happened yet. Whereas `blkid` will probe the block device directly.

Anyways, that's just papering over the issue. The busybox implementations of several utilities (especially the util-linux ones) are very minimal. If you enable the busybox module, it will happily symlink any applet that's included by busybox. The only way to replace the busybox symlinks with the real thing is to manually delete the symlink and install again. If you enable the busybox module and you get a toy program, there's little you can do about it.

**Comment by dbnicholson** [comment]
> @dbnicholson this is good background, but we need to narrow this down for actionable improvements that is in scope for the dracut project to improve (and not the busybox project).

Of course. [Here](https://github.com/dbnicholson/dracut/commit/b9e7b6227c9a2c1e7a69676afb87ea1a6f904234) is a patch I wrote yesterday to workaround our specific breakage. That simply hardcodes `blkid` (with potential for later additions like `modprobe`) to be skipped. I'm confident you don't want busybox's `blkid` as it would break every usage within dracut except for the single bare `blkid` in `rdsosreport`.

What I suggested earlier was providing configuration options such as `busyboxapplets`/`omit_busyboxapplets`. In the busybox module setup you'd either use the list in `busyboxapplets`, start with `busybox --list` and trim out `omit_busyboxapplets`, or just use the full `busybox --list` like is done now.

> > If you enable the busybox module and you get a toy program, there's little you can do about it.
> 
> Would --include help here as that would overwrite the symlink ? `dracut --include SOURCE TARGET`

I don't think so. The `--include` processing happens after the modules and uses `inst` for files. That silently skips if the target already exists.

> > In our case we have a dracut module
> 
> Depending on the "ordering of your dracut module" are you not able to manually opt in to some util-linux binaries in your dracut module ?

We'd have to order our module before `05busybox`, which is possible but not what I'd expect to do if I was just writing a regular dracut module. However, at present I'm pretty sure it wouldn't work since the busybox module doesn't check for a target that already exists and then forcefully overwrites with `ln_r`.

**Comment by devkontrol** [comment]
Continued here - https://github.com/dracut-ng/dracut/issues/2437 (corss-linking)
